### PR TITLE
chore: add api versioning changes

### PR DESCRIPTION
**Pull request by Abhinandan-Purkait**
- Move Registration proto to v1 package.
- Allow dataplane to send supported version on regisration.
- Allow dataplane to send registration info along with liveness.

Signed-off-by: Abhinandan Purkait <abhinandan.purkait@mayadata.io>

### CODE REVIEW

**Comment by Abhinandan-Purkait** [comment]
bors try

File: io-engine/src/core/env.rs [hunk 1]
```diff
@@ -75,10 +80,10 @@ fn parse_mb(src: &str) -> Result<i32, String> {
 
 #[derive(Debug, Clone, StructOpt)]
 #[structopt(
-    name = package_description!(),
-    about = "Containerized Attached Storage (CAS) for k8s",
-    version = version_info_str!(),
-    setting(structopt::clap::AppSettings::ColoredHelp)
+name = package_description ! (),
```
Comment (tiagolobocastro) [inline io-engine/src/core/env.rs]: The previous seemed better formatted?

File: io-engine/src/core/env.rs [hunk 2]
```diff
@@ -141,6 +146,15 @@ pub struct MayastorCliArgs {
     #[structopt(short = "T", long = "tgt-iface", env = "NVMF_TGT_IFACE")]
     /// NVMF target interface (ip, mac, name or subnet).
     pub nvmf_tgt_interface: Option<String>,
+    /// api Version
+    #[structopt(
+        short = "x",
```
Comment (tiagolobocastro) [inline io-engine/src/core/env.rs]: out of curiosity, why x?

File: io-engine/src/subsys/registration/registration_grpc.rs [hunk 3]
```diff
@@ -19,6 +19,27 @@ const HTTP_KEEP_ALIVE_INTERVAL: Duration = Duration::from_secs(10);
...
+    V0,
+    /// V1 version of api
+    V1,
+}
+
+impl FromStr for ApiVersion {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
```
Comment (tiagolobocastro) [inline io-engine/src/subsys/registration/registration_grpc.rs]: s.to_lower() and then can simply check "v0" and "v1" ?

File: io-engine/src/subsys/registration/registration_grpc.rs [hunk 4]
```diff
@@ -19,6 +19,27 @@ const HTTP_KEEP_ALIVE_INTERVAL: Duration = Duration::from_secs(10);
...
+            "v0" | "V0" => Ok(Self::V0),
+            "v1" | "V1" => Ok(Self::V1),
+            _ => Err("The version entered is not supported".to_string()),
```
Comment (tiagolobocastro) [inline io-engine/src/subsys/registration/registration_grpc.rs]: include the version entered?

File: io-engine/src/grpc/server.rs [hunk 5]
```diff
@@ -30,35 +35,99 @@ impl MayastorGrpcServer {
...
-                address.clone(),
-            )))
-            .add_service(v1::pool::PoolRpcServer::new(PoolService::new()))
-            .add_service(v1::replica::ReplicaRpcServer::new(
-                ReplicaService::new(),
-            ))
-            .add_service(v1::host::HostRpcServer::new(HostService::new()))
-            .add_service(v1::nexus::NexusRpcServer::new(NexusService::new()))
-            .serve(endpoint);
+        if api_versions.contains(&ApiVersion::V1)
```
Comment (tiagolobocastro) [inline io-engine/src/grpc/server.rs]: What about something like this:
```rust
        let enable_v0 = api_versions.contains(&ApiVersion::V0).then(|| true);
        let enable_v1 = api_versions.contains(&ApiVersion::V1).then(|| true);
        Server::builder()
            .add_optional_service(
                enable_v1
                    .map(|_| v1::bdev::BdevRpcServer::new(BdevService::new())),
            )
            .add_optional_service(enable_v0.map(|_| {
                MayastorRpcServer::new(MayastorSvc::new(Duration::from_millis(
                    4,
                )))
            }))
            .serve(endpoint)
            .await
```

File: io-engine/src/core/env.rs (hunk 1 above)
Comment (Abhinandan-Purkait) [reply io-engine/src/core/env.rs]: Reverting back, had used IDE formatting.

File: io-engine/src/core/env.rs (hunk 2 above)
Comment (Abhinandan-Purkait) [reply io-engine/src/core/env.rs]: Everything else seemed to be occupied. Any suggestions?

File: io-engine/src/subsys/registration/registration_grpc.rs (hunk 3 above)
Comment (Abhinandan-Purkait) [reply io-engine/src/subsys/registration/registration_grpc.rs]: Ahh right.

File: io-engine/src/subsys/registration/registration_grpc.rs (hunk 4 above)
Comment (Abhinandan-Purkait) [reply io-engine/src/subsys/registration/registration_grpc.rs]: Done.

File: io-engine/src/grpc/server.rs (hunk 5 above)
Comment (Abhinandan-Purkait) [reply io-engine/src/grpc/server.rs]: Done. Thanks a lot. This makes it much cleaner.

File: io-engine/src/core/env.rs (hunk 2 above)
Comment (tiagolobocastro) [reply io-engine/src/core/env.rs]: Not sure, do we even need a short form?

Comment (Abhinandan-Purkait) [reply io-engine/src/core/env.rs]: Probably not. Will remove it.

**Comment by Abhinandan-Purkait** [comment]
bors try

**Comment by Abhinandan-Purkait** [comment]
bors try

**Comment by Abhinandan-Purkait** [comment]
bors try

**Comment by Abhinandan-Purkait** [comment]
bors try

**Review by dsavitskiy — APPROVED** [review]

**Review by tiagolobocastro — APPROVED** [review]

**Comment by Abhinandan-Purkait** [comment]
bors try

**Comment by Abhinandan-Purkait** [comment]
bors try

**Comment by Abhinandan-Purkait** [comment]
bors merge

**Pull request merged**